Make Her quack more like ActiveRecord, ensure we will get useful error bucketing in Errbit

ceslami opened the pull request.

jmileham commented on .idea/compiler.xml:
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
gitignore all dis

jmileham commented on lib/her/model/http.rb:
```diff
@@ -54,11 +54,15 @@ def use_api(value = nil)
+          parsed_data = request[:parsed_data]
+          response = request[:response]
+
+          raise Her::Errors::RemoteServerError.new(params) if response.status >= 500
```
This is probably too granular for proper bucketing.  I think you'll want the URL that was being hit, the action that was being sent, and the status code returned, but not the parameters sent.

jmileham commented on lib/her/model/relation.rb:
```diff
@@ -99,7 +99,7 @@ def find(*ids)
               resource.run_callbacks :find
             else
-              return nil
+              raise Her::Errors::RecordNotFound.new(params.to_s)
```
Again, I think the params are going to break the bucketing too hard here.  Here we care about the class name of the thing that couldn't be found, and maybe the URL we were looking for it at (without any identifiers though), but we'll get the rest of the details via context from Errbit.  Either it'll be an end user thing and we'll get the rack context, or it'll be a DJ thing, and there'll be a stuck job to investigate.

ceslami commented:
@jmileham @samandmoore @reustmd I think this should give us the coverage/behavior we need.

ceslami commented:
@jmileham @samandmoore @reustmd Scratch that, I need a second.

ceslami commented:
@jmileham @reustmd @samandmoore I'm looking for a first round of feedback on this before I go ahead and make the (sweeping) change to look for errors in `errorMessages` instead of errors. Thanks!

ceslami commented:
@jmileham @reustmd @samandmoore My update for `errors` => `errorMessages` is done, but i'm gonna wait to push it because it will really muddy the waters

jmileham commented on lib/her/errors.rb:
```diff
@@ -20,11 +20,20 @@ class ResourceInvalid < StandardError
 
-    class RemoteResourceError < StandardError
+    class RemoteServerError < StandardError
+      def initialize(klass, code)
```
I'm still inclined to think that making these error classes have a non-standard initialization signature is overkill.

ceslami commented:
@jmileham done!

reustmd commented on lib/her/model/orm.rb:
```diff
@@ -75,7 +75,7 @@ def destroy
         run_callbacks :destroy do
           self.class.request(:_method => method, :_path => request_path) do |parsed_data, response|
-            raise Her::Errors::RemoteResourceError if response.status >= 400
+            raise Her::Errors::RecordNotFound.new("Request against #{self.class} returned a #{response.status}") if response.status >= 400
```
should this be a bit more structured? For this plaid MFA stuff, we're going to have to inspect and react to a fairly specific error.

For example, we get this response from plaid sometimes:

402
{
  "code": 1215,
  "message": "mfa reset",
  "resolve": "MFA access has changed or this application's access has been revoked. Submit a PATCH call to resolve."
}

reustmd replied:
More specifically, I'm wondering if the exception should have specific properties for `response_body` and/or `http_code`

ceslami replied:
So its got HTTP code, but I think we decided that response_body would be a bit too specific for Errbit to bucket effectively. Plus, there should be other ways for us to diagnose a specific error like you mentioned, even if all the details don't appear in Errbit. @jmileham does that sound right?

reustmd replied:
yeah, i don't mean for errbit, i mean for consumer code to capture an exception and react properly. plaid uses lots of diff 40x responses with relevant body content.

ceslami replied:
Ah, I see. I think we can address these as use-cases emerge. `RecordNotFound` definitely doesn't sound right for a 402, but we know that we want to throw for the 400 space. 

I wonder what the best approach here is. Should we special-case out some 400 space codes (409, 404, 402), and then have a catch-all for the rest?

reustmd replied:
I think it's going to vary too much, api-to-api to be semantic. I just think a general exception, with structured access, `exception.http_status_code` is :cool:

jmileham replied:
Yeah, there's a tension here between Errbit bucketing, general diagnostic ability, and security (e.g. API responses could have arbitrarily sensitive content).  One problem here is that Her doesn't/shouldn't have logger awareness, or I'd maybe suggest that we use Rails.logger to drop extra metadata.  Perhaps we can drop in custom her middleware that rescues exceptions raised here and logs appropriate diagnostic metadata before reraising.  That allows us granular control over what diagnostics we log on a per-service or even per-endpoint basis.

So I guess I'm in favor of making these exceptions semantic about what kind of HTTP status code they returned so we can then catch them in app-specific middleware and give ourselves strong app-specific diagnostics.

samandmoore replied:
not sure if we lost a little bit of focus here, but this specific convo is happening on the code within `destroy` which should not, generally, include error messages that we can actually act upon, right? even from external partners, i would imagine those to use a POST for "destroying" resources.

jmileham replied:
I think this has become a general conversation about error states and how to divide the concerns of error behavior and diagnostics, securely, in an API-specific way.  Destroy operations are probably not the most likely to have recoverable (or confusing) reasons behind their failures.

samandmoore replied:
>  Perhaps we can drop in custom her middleware that rescues exceptions raised here and logs appropriate diagnostic metadata before reraising.

i don't think you can insert her middleware at a point _after_ the "destroy" method and the like are called. the middlewares are used only in faraday land, so they influence only how the data is passed _into_ the her ORM abstraction.

samandmoore replied:
if we want to use a middleware-based solution, then we have to go deeper down in this gem's code and insert this raise-y behavior there.

jmileham replied:
Yeah, maybe middleware isn't right, but basically diagnostics should be an
app layer concern, and if we raise activerecord-y things (RecordNotFound,
RecordInvalid), the app layer should be able to layer in diagnostics
somehow.

On Tue, Jun 23, 2015 at 12:08 PM, Sam Moore <notifications@github.com>
wrote:

> In lib/her/model/orm.rb
> <https://github.com/Betterment/her/pull/4#discussion_r33059041>:
>
> > @@ -75,7 +75,7 @@ def destroy
> >          method = self.class.method_for(:destroy)
> >          run_callbacks :destroy do
> >            self.class.request(:_method => method, :_path => request_path) do |parsed_data, response|
> > -            raise Her::Errors::RemoteResourceError if response.status >= 400
> > +            raise Her::Errors::RecordNotFound.new("Request against #{self.class} returned a #{response.status}") if response.status >= 400
>
> if we want to use a middleware-based solution, then we have to go deeper
> down in this gem's code and insert this raise-y behavior there.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Betterment/her/pull/4/files#r33059041>.
>

ceslami commented:
@jmileham @samandmoore @reustmd bump!

ceslami commented:
@jmileham @samandmoore @reustmd I think this correctly reflects the design we discussed earlier. Dropped handling of a 409 and added in a catch-all exception for 400-space status codes.

cc/ @LucasSantoro in case you're interested

jmileham commented on lib/her/errors.rb:
```diff
@@ -15,16 +15,29 @@ class AssociationUnknownError < StandardError
+          when 422
+            RecordInvalid
+          else
+            StandardClientError
```
Starting to think that `ClientError` should be called `ResponseError`, and should itself inherit from `StandardError`.  Then RemoteServerError, RecordNotFound and RecordInvalid should inherit from that?

ceslami replied:
After giving it a day, I really prefer `ClientError`. It isn't clear to me that `ResponseError` means 400-space specifically, and the HTTP Spec refers to the 400-space as "Client Errors."

That said, I like the inheritance idea!

reustmd replied:
`ClientError` feels fine to me.

ceslami commented:
Alright guys, lookin' good feelin' good.

jmileham commented on lib/her/errors.rb:
```diff
@@ -15,16 +15,29 @@ class AssociationUnknownError < StandardError
+    class RemoteServerError < StandardError
+    end
+
+    class RecordNotFound < StandardError
```
If you're making the point that ClientError is so named because it models 4xx (which I think is maybe not super useful because these exceptions are about hiding HTTP semantics and replacing with AR-like semantics), then at least RecordNotFound and RecordInvalid should inherit from ClientError.  But I'd actually argue, as I did before, that it is more valuable for all of the her errors to inherit from a single base class that doesn't just represent 4xx, but represents "the server didn't do what we wanted here."  That could even include a 302 redirect we weren't expecting.  Then our user land code can catch any such error consistently.  It's possible that adding ClientError as a middle inheritence tier could be valuable, but I think we'd add it when a use case had proven itself.  It would only be useful if app code cared to catch all sub genres of 4xx, but nothing else, and as I was saying, the point of Her is to replace HTTP semantics with AR semantics (where not found and invalid are really not all that much more semantically similar than a generic exception or unexpected redirect), so that sounds like it'd be of limited value.

jmileham commented:
Love it.  LGTM - second eyeballs given the contention here?

reustmd commented:
lgtm
